src/lib.rs

```diff
@@ -82,26 +82,93 @@ impl Gtid {
 
     /// Add an interval but does not check assume interval is correctly formed
     fn add_interval_unchecked(&mut self, interval: &(u64, u64)) {
-        let mut interval = *interval;
+        let interval = *interval;
 
-        // TODO: Do it in place with a filter.
-        let mut new: Vec<(u64, u64)> = Vec::with_capacity(self.intervals.len());
-        for current in self.intervals.iter() {
-            if interval.0 == current.1 {
-                interval = (current.0, interval.1);
-                continue;
+        // self.intervals was empty
+        if self.intervals.is_empty() {
+            self.intervals.push(interval);
+            return;
+        }
+        if let Some(&first) = self.intervals.first() {
+            // if interval is strictly before intervals
+            if interval.1 < first.0 {
+                self.intervals.insert(0, interval);
+                return;
             }
 
-            if interval.1 == current.0 {
-                interval = (interval.0, current.1);
-                continue;
+            // if interval merge before interval
+            if interval.1 == first.0 {
+                // unwrapping is ok as intervals is not empty
+
+                let first = self.intervals.first_mut().unwrap();
+                *first = (interval.0, first.1);
```

Comment:
nit: only updating the left bound, so:

```suggestion
                first.0 = interval.0;
```